src/environment.py

```diff
@@ -0,0 +1,298 @@
+"""Generate container app environment based on charm configuration."""
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import os
+from urllib.parse import urlparse
+
+# Mattermost's default port, and what we expect the image to use
+CONTAINER_PORT = 8065
+# Default port, enforced via envConfig to prevent operator error
+METRICS_PORT = 8067
+DATABASE_NAME = "mattermost"
+REQUIRED_S3_SETTINGS = ["s3_bucket", "s3_region", "s3_access_key_id", "s3_secret_access_key"]
+REQUIRED_SETTINGS = ["mattermost_image_path"]
+REQUIRED_SSO_SETTINGS = ["license", "site_url"]
+SAML_IDP_CRT = "saml-idp.crt"
+
+
+def _env_for_canonical_defaults(config: dict) -> dict:
+    """Return various Mattermost settings particular to Canonical's deployment.
+
+    These settings may be less generally useful, and so they are controlled here as a unit.
+
+    Args:
+        config: dict of the charm's configuration
+
+    Returns:
+        dict of various Mattermost settings particular to Canonical's deployment.
+    """
+    if not config["use_canonical_defaults"]:
+        return {}
+    return {
+        # If this is off, users can't turn it on themselves.
+        "MM_SERVICESETTINGS_CLOSEUNUSEDDIRECTMESSAGES": "true",
+        # This allows Matterhorn to use emoji and react to messages.
+        "MM_SERVICESETTINGS_ENABLECUSTOMEMOJI": "true",
+        # If this is off, users can't turn it on themselves.
+        "MM_SERVICESETTINGS_ENABLELINKPREVIEWS": "true",
+        # Matterhorn recommends the use of Personal Access Tokens.
+        "MM_SERVICESETTINGS_ENABLEUSERACCESSTOKENS": "true",
+        # We'll use one large team.  Create and invite are
+        # disabled in the System Scheme, found in the Permissions
+        # section of the System Console.
+        "MM_TEAMSETTINGS_MAXUSERSPERTEAM": "1000",
+    }
+
+
+def _env_for_clustering(config: dict, app_name: str) -> dict:
+    """Return clustering settings, varying the cluster name on the application name.
+
+    This is done so that blue/green deployments in the same model won't talk to each other.
+
+    Args:
+        config: dict of the charm's configuration
+        app_name: name of the charm's app
+
+    Returns:
+        dict of clustering settings, varying the cluster name on the application name.
+    """
+    if not config.get("clustering"):
+        return {}
+    return {
+        "MM_CLUSTERSETTINGS_ENABLE": "true",
+        # https://juju.is/docs/sdk/charm-environment-variables
```

Comment:
What is this comment? Is it still relevant?

Reply:
I had difficulties finding this page so I thought it was interesting to have it as a comment. I can remove it.

Reply:
Perhaps add some context why it is relevant?

Reply:
Context added